File: backend/jasprchain/consensus/block.py
"""Block Structure for JasprChain
Mirrors Rust consensus block types
"""
from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any
from datetime import datetime, timezone
import json
import hashlib


@dataclass
class BlockHeader:
    """Block header containing consensus-critical data
    
    Mirrors Rust struct for easy porting
    """
    height: int
    previous_hash: str
    timestamp: int  # Unix timestamp in milliseconds
    proposer: str  # Validator address
    state_root: str  # Sparse Merkle Tree root
    transactions_root: str  # Merkle root of transactions
    receipts_root: str  # Merkle root of execution receipts
    
    # Consensus fields
    committee_signatures: str = ""  # Aggregated BLS signature
    attestation_count: int = 0  # Number of validators who attested
    
    def hash(self) -> str:
        """Calculate block hash"""
        data = json.dumps({
            'height': self.height,
            'previous_hash': self.previous_hash,
            'timestamp': self.timestamp,
            'proposer': self.proposer,
            'state_root': self.state_root,
            'transactions_root': self.transactions_root,
            'receipts_root': self.receipts_root
        }, sort_keys=True).encode()
        return hashlib.sha256(data).hexdigest()
    
    def to_dict(self) -> dict:
        return {
            'height': self.height,
            'previous_hash': self.previous_hash,
            'timestamp': self.timestamp,
            'proposer': self.proposer,
            'state_root': self.state_root,
            'transactions_root': self.transactions_root,
            'receipts_root': self.receipts_root,
            'committee_signatures': self.committee_signatures,
            'attestation_count': self.attestation_count
        }


@dataclass
class Block:
    """Complete block with header and transactions
    
    Mirrors Rust Block struct
    """
    header: BlockHeader
    transactions: List[Dict[str, Any]] = field(default_factory=list)
    
    # Execution results
    receipts: List[Dict[str, Any]] = field(default_factory=list)
    
    # Finality status
    finalized: bool = False
    finality_time_ms: Optional[int] = None
    
    @property
    def hash(self) -> str:
        return self.header.hash()
    
    @property
    def height(self) -> int:
        return self.header.height
    
    def to_dict(self) -> dict:
        return {
            'hash': self.hash,
            'header': self.header.to_dict(),
            'transactions': self.transactions,
            'transaction_count': len(self.transactions),
            'receipts': self.receipts,
            'finalized': self.finalized,
            'finality_time_ms': self.finality_time_ms
        }
    
    def to_summary(self) -> dict:
        """Compact summary for API responses"""
        return {
            'hash': self.hash,
            'height': self.height,
            'timestamp': self.header.timestamp,
            'proposer': self.header.proposer,
            'tx_count': len(self.transactions),
            'finalized': self.finalized,
            'state_root': self.header.state_root[:16] + '...'
        }


def create_genesis_block() -> Block:
    """Create the genesis block for JasprChain"""
    now = int(datetime.now(timezone.utc).timestamp() * 1000)
    
    header = BlockHeader(
        height=0,
        previous_hash="0" * 64,
        timestamp=now,
        proposer="jaspr1genesis",
        state_root=hashlib.sha256(b'genesis_state').hexdigest(),
        transactions_root=hashlib.sha256(b'').hexdigest(),
        receipts_root=hashlib.sha256(b'').hexdigest(),
        committee_signatures="",
        attestation_count=0
    )
    
    genesis = Block(
        header=header,
        transactions=[],
        receipts=[],
        finalized=True,
        finality_time_ms=0
    )
    
    return genesis

File: rust-core/jasprchain/src/consensus/block.rs
//! Block structures for JasprChain

use serde::{Deserialize, Serialize};
use sha2::{Sha256, Digest};
use std::time::{SystemTime, UNIX_EPOCH};

/// Block header containing consensus-critical data
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BlockHeader {
    pub height: u64,
    pub previous_hash: String,
    pub timestamp: u64,
    pub proposer: String,
    pub state_root: String,
    pub transactions_root: String,
    pub receipts_root: String,
    pub committee_signatures: Vec<u8>,
    pub attestation_count: u32,
}

impl BlockHeader {
    pub fn hash(&self) -> String {
        let data = serde_json::to_string(&self).unwrap_or_default();
        let mut hasher = Sha256::new();
        hasher.update(data.as_bytes());
        hex::encode(hasher.finalize())
    }
}

/// Complete block with header and transactions
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Block {
    pub header: BlockHeader,
    pub transactions: Vec<serde_json::Value>,
    pub receipts: Vec<serde_json::Value>,
    pub finalized: bool,
    pub finality_time_ms: Option<u64>,
}

impl Block {
    /// Create genesis block
    pub fn genesis() -> Self {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        
        let header = BlockHeader {
            height: 0,
            previous_hash: "0".repeat(64),
            timestamp: now,
            proposer: "jaspr1genesis".to_string(),
            state_root: Self::hash_bytes(b"genesis_state"),
            transactions_root: Self::hash_bytes(b""),
            receipts_root: Self::hash_bytes(b""),
            committee_signatures: Vec::new(),
            attestation_count: 0,
        };
        
        Self {
            header,
            transactions: Vec::new(),
            receipts: Vec::new(),
            finalized: true,
            finality_time_ms: Some(0),
        }
    }
    
    pub fn hash(&self) -> String {
        self.header.hash()
    }
    
    fn hash_bytes(data: &[u8]) -> String {
        let mut hasher = Sha256::new();
        hasher.update(data);
        hex::encode(hasher.finalize())
    }
}
